File: research/integrity_tranche5b_research_lab_freshness.js
#!/usr/bin/env node
'use strict';
const fs=require('fs'),path=require('path'),vm=require('vm'),assert=require('assert');
const ROOT=path.resolve(__dirname,'..');
const expected=JSON.parse(fs.readFileSync(path.join(ROOT,'config/research-lab-ux-contract.json'),'utf8'));
const elements=new Map();
const document={
 readyState:'complete',
 getElementById:id=>elements.get(id)||null,
 addEventListener:()=>{}
};
const context={window:{addEventListener:()=>{},FIE_M5:{getCurrentBundle:()=>null}},document,console,Date,Object,Number,String,Math};
context.window.window=context.window;context.window.document=document;
vm.createContext(context);
vm.runInContext(fs.readFileSync(path.join(ROOT,'app/core/research-lab-ux.js'),'utf8'),context);
const api=context.window.FIEFreshness,ux=context.window.FIEResearchLabUX;
assert(api&&ux,'Research/Lab UX APIs missing');
assert.deepStrictEqual(JSON.parse(JSON.stringify(api.contract)),expected,'browser contract must exactly mirror JSON contract');
assert(Object.isFrozen(api.contract)&&Object.isFrozen(api.contract.navigation_groups),'contract must be immutable');
const now=Date.parse('2026-09-04T12:00:00Z');
assert.strictEqual(api.describe({asOf:'2026-09-04T10:00:00Z',maxAgeHours:8,now}).state,'current');
assert.strictEqual(api.describe({asOf:'2026-09-04T05:30:00Z',maxAgeHours:8,now}).state,'aging');
assert.strictEqual(api.describe({asOf:'2026-09-03T12:00:00Z',maxAgeHours:8,now}).state,'stale');
assert.strictEqual(api.describe({asOf:'2026-09-04T10:00:00Z',available:false,maxAgeHours:8,now}).state,'unavailable','availability must fail closed before timestamp freshness');
assert.strictEqual(api.describe({asOf:'not-a-time',available:true,maxAgeHours:8,now}).state,'unknown');
assert.strictEqual(api.describe({asOf:'2026-09-05T12:00:00Z',available:true,maxAgeHours:8,now}).state,'unknown','future timestamps cannot be current');
assert.strictEqual(api.weeklyLabel({asOf:null,available:false}),'Unavailable','missing week/season must not become Season 0');
assert.strictEqual(api.isFresh({asOf:'2026-09-04T05:30:00Z',maxAgeHours:8,now}),true);
assert.strictEqual(api.isFresh({asOf:'2026-09-03T12:00:00Z',maxAgeHours:8,now}),false);
assert(api.weeklyLabel({season:2026,week:1,asOf:'2026-09-04T10:00:00Z',maxAgeHours:8,now}).startsWith('Season 2026 · Week 1 · Current'));
for(const key of ['score','rank','recommendation','activation'])assert(!Object.prototype.hasOwnProperty.call(api.describe({}),key),`display presenter must not emit ${key}`);
const index=fs.readFileSync(path.join(ROOT,'index.html'),'utf8'),report=fs.readFileSync(path.join(ROOT,'app/research-report-ui.js'),'utf8'),runtime=fs.readFileSync(path.join(ROOT,'app/runtime-foundation.js'),'utf8');
assert(index.includes('data-fie-research-report')&&report.includes('[data-fie-research-report]'),'League Research Report entry point is not integrated');
assert(index.indexOf('app/core/research-lab-ux.js')<index.indexOf('app/runtime-foundation.js'),'freshness presenter must load before runtime diagnostics');
assert(index.includes("window.FIEFreshness?.weeklyLabel")&&index.includes("window.FIEFreshness?.isFresh"),'M5/M6 do not use shared freshness');
assert(report.includes('window.FIEFreshness?.timestampLabel')&&runtime.includes('window.FIEFreshness?.timestampLabel'),'report/runtime freshness not unified');
console.log('PASS Tranche 5B grouped Research/Lab IA and display-only freshness contract');

File: research/integrity_tranche5b_research_lab_freshness.py
#!/usr/bin/env python3
"""Tranche 5B Research/Lab IA and freshness characterization."""
from __future__ import annotations

import argparse
import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
SHELL = (ROOT / "index.html").read_text(encoding="utf-8")
REPORT = (ROOT / "app/research-report-ui.js").read_text(encoding="utf-8")
EVIDENCE = (ROOT / "app/core/evidence-semantics.js").read_text(encoding="utf-8")
RUNTIME = (ROOT / "app/runtime-foundation.js").read_text(encoding="utf-8")


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--mode", choices=("baseline", "target"), default="baseline")
    args = ap.parse_args()

    lab_match = re.search(r"lab:\{title:'Lab'.*?\}\n", SHELL)
    assert lab_match, "Lab navigation contract not found"
    tabs_match = re.search(r"tabs:\[(.*?)\](?:,routes:|\})", lab_match.group(0))
    assert tabs_match, "Lab visible tabs not found"
    lab_tabs = re.findall(r"\['([^']+)','([^']+)'\]", tabs_match.group(1))
    milestone_panels = [f"research{i}Panel" if i > 1 else "researchPanel" for i in range(1, 7)]
    facts = {
        "lab_tab_count": len(lab_tabs),
        "lab_tabs": [label for _, label in lab_tabs],
        "milestone_panel_count": sum(f'id="{panel}"' in SHELL for panel in milestone_panels),
        "standalone_research_report_overlay": "fieResearchOverlay" in REPORT,
        "standalone_research_report_launcher": "fieResearchLauncher" in REPORT,
        "integrated_research_report_entry": "data-fie-research-report" in SHELL and "data-fie-research-report" in REPORT,
        "lab_overview": 'id="labOverviewPanel"' in SHELL,
        "legacy_routes_preserved": all(f"'{route}'" in lab_match.group(0) for route in ("research", "research2", "research3", "research4", "features", "model")),
        "typed_as_of_available": "function firstAsOf" in EVIDENCE and "asOf:" in EVIDENCE,
        "m5_week_only_badge": "Current ${c?.season||''} W${c?.week||'?'}" in SHELL,
        "m6_boolean_freshness": "function m6Fresh" in SHELL,
        "raw_build_timestamp": "BUILD_MANIFEST.generated_at" in RUNTIME or "BUILD_MANIFEST?.generated_at" in RUNTIME,
        "shared_freshness_presenter": "FIEFreshness" in "\n".join((SHELL, REPORT, RUNTIME)),
    }

    assert facts["milestone_panel_count"] == 6, facts
    assert facts["standalone_research_report_overlay"], facts
    assert facts["typed_as_of_available"], facts
    assert facts["m6_boolean_freshness"], facts
    assert facts["raw_build_timestamp"], facts

    if args.mode == "baseline":
        assert facts["lab_tab_count"] == 9, facts
        assert facts["standalone_research_report_launcher"], facts
        assert facts["m5_week_only_badge"], facts
        assert not facts["shared_freshness_presenter"], facts
        print("KNOWN_GAP_REPRODUCED Research/Lab navigation and freshness presentation remain fragmented")
    else:
        contract_path = ROOT / "config/research-lab-ux-contract.json"
        assert contract_path.is_file(), "target Research/Lab UX contract missing"
        contract = json.loads(contract_path.read_text(encoding="utf-8"))
        assert contract.get("schema") == "fie-research-lab-ux-v1"
        assert contract.get("navigation_groups") and contract.get("freshness_states")
        assert facts["lab_tab_count"] == 4, facts
        assert facts["lab_overview"], facts
        assert facts["integrated_research_report_entry"], facts
        assert facts["legacy_routes_preserved"], facts
        assert not facts["m5_week_only_badge"], facts
        assert facts["shared_freshness_presenter"], facts
        print("TARGET_GAP_CLOSED Research/Lab navigation grouped and freshness presentation unified")
    print(json.dumps({"mode": args.mode, **facts}, sort_keys=True))


if __name__ == "__main__":
    main()
